.github/workflows/ci.yaml

```diff
@@ -18,44 +18,131 @@ jobs:
       run: echo ${{ github.ref }}
     - name: event_name
       run: echo ${{ github.event_name }}
-  build:
-    name: Build
+  lint:
+    name: Lint
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v3
+    - uses: actions/setup-go@v2
+      with:
+        go-version: 1.17
+    - name: golangci-lint
+      uses: golangci/golangci-lint-action@v3.1.0
+      with:
+        version: v1.44.2
+  test:
+    name: Test
     runs-on: ubuntu-latest
     steps:
     - name: checkout
-      uses: actions/checkout@v2
+      uses: actions/checkout@v3
+
     - uses: actions/setup-go@v2
       with:
-        go-version: '1.17' # The Go version to download (if necessary) and use.
-    - name: ci
-      run: make ci
+        go-version: 1.17
+    - id: go-cache-paths
+      shell: bash
+      run: |
+        echo "::set-output name=go-build::$(go env GOCACHE)"
+        echo "::set-output name=go-mod::$(go env GOMODCACHE)"
+    - name: Go Mod Cache
+      uses: actions/cache@v2
+      with:
+        path: ${{ steps.go-cache-paths.outputs.go-mod }}
+        key: ${{ runner.os }}-go-mod-${{ hashFiles('**/go.sum') }}
+        restore-keys: |
+          ${{ runner.os }}-go-mod-
+    - name: Go Build Cache
+      uses: actions/cache@v2
+      with:
+        path: ${{ steps.go-cache-paths.outputs.go-build }}
+        key: ${{ runner.os }}-go-build-${{ hashFiles('**/go.sum') }}
+        restore-keys: |
+          ${{ runner.os }}-go-build-
+
+    - name: test
+      run: make race
+
+  image:
```

Comment:
Separate job for building and pushing the image

Reply:
Should this not depend on the results of tests and lint? We only want to push out if build, test, lint pass. And only if this is merged to master or tagged, not PRs on other branches.

Reply:
Currently this replaces the previous build job, and pushing is disabled for PRs and dependabot branches which will lack access to the secrets in order to push, so this will just verify that the image (and source) can be built successfully in those cases.

We can exclude branches other than `master`, but I would argue that we could benefit from pushing images for other branches for easier testing purposes, and anyone pushing to a branch on this repo could be considered trusted from a secrets perspective.

Tags will be excluded here because we are specifying branches to match against, as can be seen from the workflows being run against my fork: https://github.com/detiber/packet-ccm/actions

Reply:
It isn't just a question of what has the right to push, although that is a big issue. I also don't think we want images out there and publicly available for anything other than what has been approved and merged. People can end up using them accidentally, and there may even be ownership/responsibility/liability issues.

Let's do the following. We currently only ever push out for merged-to-master and semver-tagged. Let's keep it this way, so that this PR focuses on all of the goodness we gain without changing behaviour. Then, if you want, open a separate PR that just pushes those out, and we can focus the discussion there.

Does that make sense?